### PR TITLE
bpo-44156: Make cached string constants in compile.c subinterpreter compatible

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
https://bugs.python.org/issue44156

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
@vstinner, please take a look if you have the time. Thank you!

**Comment by Fidget-Spinner** [comment]
Some tests are failing because of GH action outages https://www.githubstatus.com/incidents/zbpwygxwb3gw.

**Comment by vstinner** [comment]
Please create an issue for this change. You can a "meta" issue to make modules static variables compatible with sub-interpreters, but it's better to create a specific issue. For example, only replace "string constants".

See for example https://bugs.python.org/issue40521 but please don't reuse this one.

**Comment by Fidget-Spinner** [comment]
> Please create an issue for this change. You can a "meta" issue to make modules static variables compatible with sub-interpreters, but it's better to create a specific issue. For example, only replace "string constants".
> 
> See for example https://bugs.python.org/issue40521 but please don't reuse this one.

Thanks Victor. I've followed your recommendations.

**Comment by Fidget-Spinner** [comment]
I just noticed I need to regenerate the blurb. I'll do that soon.

File: Python/compile.c [hunk 1]
```diff
@@ -2250,11 +2246,6 @@ compiler_visit_annotations(struct compiler *c, arguments_ty args,
                                      args->kwarg->annotation, &annotations_len))
         return 0;
 
-    if (!return_str) {
-        return_str = PyUnicode_InternFromString("return");
-        if (!return_str)
-            return 0;
-    }
```
Comment (vstinner) [inline Python/compile.c:2257]: I would prefer to only define return_str here.

File: Python/compile.c [hunk 2]
```diff
@@ -5018,11 +5005,10 @@ compiler_setcomp(struct compiler *c, expr_ty e)
 static int
 compiler_dictcomp(struct compiler *c, expr_ty e)
 {
-    static identifier name;
-    if (!name) {
-        name = PyUnicode_InternFromString("<dictcomp>");
-        if (!name)
-            return 0;
+    _Py_static_string(PyId_name, "<dictcomp>");
```
Comment (vstinner) [inline Python/compile.c]: Can you please rename the variable to PyId_dictcomp?

File: Python/compile.c [hunk 3]
```diff
@@ -5002,11 +4990,10 @@ compiler_listcomp(struct compiler *c, expr_ty e)
 static int
 compiler_setcomp(struct compiler *c, expr_ty e)
 {
-    static identifier name;
-    if (!name) {
-        name = PyUnicode_InternFromString("<setcomp>");
-        if (!name)
-            return 0;
+    _Py_static_string(PyId_name, "<setcomp>");
```
Comment (vstinner) [inline Python/compile.c]: Can you please rename the variable to PyId_setcomp?

File: Python/compile.c [hunk 4]
```diff
@@ -4987,11 +4976,10 @@ compiler_genexp(struct compiler *c, expr_ty e)
 static int
 compiler_listcomp(struct compiler *c, expr_ty e)
 {
-    static identifier name;
-    if (!name) {
-        name = PyUnicode_InternFromString("<listcomp>");
-        if (!name)
-            return 0;
+    _Py_static_string(PyId_name, "<listcomp>");
```
Comment (vstinner) [inline Python/compile.c]: Rename the variable to PyId_listcomp.

File: Python/compile.c [hunk 5]
```diff
@@ -4972,11 +4962,10 @@ compiler_comprehension(struct compiler *c, expr_ty e, int type,
 static int
 compiler_genexp(struct compiler *c, expr_ty e)
 {
-    static identifier name;
-    if (!name) {
-        name = PyUnicode_InternFromString("<genexpr>");
-        if (!name)
-            return 0;
+    _Py_static_string(PyId_name, "<genexpr>");
+    identifier name = _PyUnicode_FromId(&PyId_name); /* borrowed ref */
```
Comment (vstinner) [inline Python/compile.c]: Rename the variable to PyId_genexpr.

File: Python/compile.c [hunk 6]
```diff
@@ -2799,18 +2790,18 @@ compiler_lambda(struct compiler *c, expr_ty e)
...
     assert(e->kind == Lambda_kind);
 
     if (!compiler_check_debug_args(c, args))
         return 0;
 
-    if (!name) {
-        name = PyUnicode_InternFromString("<lambda>");
-        if (!name)
-            return 0;
+    _Py_static_string(PyId_name, "<lambda>");
```
Comment (vstinner) [inline Python/compile.c]: Rename the variable to PyId_lambda.

File: Python/compile.c [hunk 7]
```diff
@@ -2232,7 +2224,11 @@ compiler_visit_annotations(struct compiler *c, arguments_ty args,
 
        Return 0 on error, -1 if no annotations pushed, 1 if a annotations is pushed.
        */
-    static identifier return_str;
+    _Py_static_string(PyId_return, "return");
```
Comment (vstinner) [inline Python/compile.c]: You can use _Py_IDENTIFIER(return) macro.

**Comment by vstinner** [comment]
Merged, thanks.

**Comment by Fidget-Spinner** [comment]
@vstinner Thank you for taking the time to review this and make the code more readable. I appreciate it :).

**Pull request merged**